### PR TITLE
[sp-185] filter by capacity and travel date & [sp-182] add controller

### PR DESCRIPTION
**Pull request by shawn-cx-li**
https://aussiecommerce.atlassian.net/browse/SP-185

### CODE REVIEW

File: src/context/search.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { IOffer, IPackage, IPackageCapacity, ITravelDate } from "../types";
```
Comment (joshuapaling) [inline src/context/search.ts:1]: The name of this file - `src/context/search.ts` is very generic. We're doing a lot of different types of searching in this repo so I think this should be given a more specific name like `filterLeOffersByDatesAndCapacity.ts` or something.

**Review by joshuapaling — CHANGES_REQUESTED** [review]
We've introduced a `/context` folder here, and now we've got:
`/lib/search.ts` and 
`/context/search.ts`

And, the only place that `/lib/search` is used is from `/context/search`. To me that feels confusing. Those files all belong together. I'd be in favour of an approach similar to the `/lib/augmentPlaceResults` folder, where everything related to `searching LE offers` sits under the one folder.

What's the difference between "lib" and "context" here? We should be clear on the distinction so we can decide what goes where, rather than them both becoming "two places where files accumulate".

Can have a call about this if it's easier.

**location search**

Maybe it's something we want to tackle in a separate PR, and the ticket might not have been clear enough here, but when we search by `location`, it won't be a string like `australia`. It'll be a google place ID, like `ChIJ_8h84N9G0i0R0P2CyvsLAwU`, and later, possibly could also be a "popular region" that we've defined (which is basically a list of Google Place IDs that make up that region).

That's probably more than we want to tackle in this PR, so happy to have a further discussion about that later, and only deal with everything above **location search** in this PR.

**Comment by shawn-cx-li** [comment]
> We've introduced a `/context` folder here, and now we've got:
> `/lib/search.ts` and
> `/context/search.ts`
> 
> And, the only place that `/lib/search` is used is from `/context/search`. To me that feels confusing. Those files all belong together. I'd be in favour of an approach similar to the `/lib/augmentPlaceResults` folder, where everything related to `searching LE offers` sits under the one folder.
> 
> What's the difference between "lib" and "context" here? We should be clear on the distinction so we can decide what goes where, rather than them both becoming "two places where files accumulate".
> 
> Can have a call about this if it's easier.
> 
> **location search**
> 
> Maybe it's something we want to tackle in a separate PR, and the ticket might not have been clear enough here, but when we search by `location`, it won't be a string like `australia`. It'll be a google place ID, like `ChIJ_8h84N9G0i0R0P2CyvsLAwU`, and later, possibly could also be a "popular region" that we've defined (which is basically a list of Google Place IDs that make up that region).
> 
> That's probably more than we want to tackle in this PR, so happy to have a further discussion about that later, and only deal with everything above **location search** in this PR.

### folder
To me the `lib` folder should have functions that serves single functionalities, such as `filterByCapacity`, `filterByAvailableDates`; while `context` is something serves a business logic, such as `searchLEDeals` in which we can do sequntial operations like: 1) find all related keys by location; 2) pull all offers by keys; 3)filter by availablility.

But yes, I agree the naming is confusing, what do you think if I change is to `/context/searchLeDeals.ts` and `/lib/searchFilters.ts`?

### location search
Yes, let's do it in another PR. But I would like to keep the current logic for a demostration purpose, in later PR we can update the module to use the new logic

**Comment by joshuapaling** [comment]
I'm happy with the name `/context/searchLeDeals.ts`. I'm also happy with having `context` folder for business logic, and `lib` for specific functions - with one caveat. If a function is only used in one place, I think it belongs alongside the code that calls it (in this case, all the code belongs under a `/context/searchLeDeals/` folder). It's very useful to be able to see it "all in once place". 

If a specific function is used in multiple call sites, from different places, sure - share it by putting it in lib.

But, it's probably not a big enough issue to hold up merging this PR, so if you disagree, you can go ahead and merge.

**Review by joshuapaling — APPROVED** [review]

File: src/lib/search.test.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    it("should always return true if checkIn or checkOut is not specified", () => {
+      const travelDateWithoutCheckInAndCheckOutDate: ITravelDate = {};
+
+      const travelDateWithOnlyCheckInDate: ITravelDate = {
+        checkIn: "2020-09-09"
+      };
+
+      const travelDateWithOnlyCheckOutDate: ITravelDate = {
+        checkOut: "2020-09-09"
+      };
```
Comment (sung-luxuryescapes) [inline src/lib/search.test.ts:53]: When would a search parameter not have both `checkIn` and `checkout`?

File: src/context/search.test.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+describe("context:search", () => {
+  describe("searchLEDeals", () => {
+    it("should filter out package_02 by capacity", async () => {
+      const inquiryCapacity: IPackageCapacity = {
+        adults: 2,
+        children: 2,
+        infants: 0
+      };
+      const inquiryTravelDate: ITravelDate = {};
```
Comment (sung-luxuryescapes) [inline src/context/search.test.ts:49]: Please see https://github.com/brandsExclusive/fn-search/pull/89/files#r473530886

**Comment by shawn-cx-li** [comment]
> I'm happy with the name `/context/searchLeDeals.ts`. I'm also happy with having `context` folder for business logic, and `lib` for specific functions - with one caveat. If a function is only used in one place, I think it belongs alongside the code that calls it (in this case, all the code belongs under a `/context/searchLeDeals/` folder). It's very useful to be able to see it "all in once place".
> 
> If a specific function is used in multiple call sites, from different places, sure - share it by putting it in lib.
> 
> But, it's probably not a big enough issue to hold up merging this PR, so if you disagree, you can go ahead and merge.

Good point, moved to `context/searchLEDeals.ts`

**Review by colins44 — APPROVED** [review]

**Pull request merged**